### PR TITLE
Improve & add components section to Editor intro

### PR DESCRIPTION
**Pull request by hyounes4560**
# Pull Request

## Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder-docs/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 4.x)
* [x] Checked your writing carefully for correct English spelling, grammar, etc
* [x] Described your changes and the motivation for them below

## Description of Changes

Updates the editor introduction section

* [x] Adds more content to expand the Editor's current intro documentation
* [x] Adds new annotated images

### CODE REVIEW

**Comment by CAM-Gerlach** [comment]
Hey @hyounes4560 ; thanks for your work so far and looking like a pretty good start! I particularly like the screenshots with the annotations and numbered sections!

I see I was tagged for review, but this PR is also still marked as a draft. Just to confirm, would you like me to review it now? If so, is there anything particular you'd like to be in or out of scope for the first round of review?

When it comes to things like reST syntax, Sphinx roles, source formatting, and other technical and style guide details, there's a few ways we can handle that, so let me know which you prefer (and we can always change it later on future PRs depending on what works best for you):

* Review suggestions on the PR, and you apply them
* Push a commit to the PR taking care of technical nits (all substantive content feedback will still be made as suggestions)
* Ignore these minor defects on this PR completely and merge with them present, and make an immediate followup PR to address them

In the past, working with our previous writer, I initially did review suggestions, but we subsequently found directly pushing a commit taking care of the small nitpicky stuff to be a lot more efficient for both of us and did that from then on, so that might be the best option, but its ultimately up to you what you want me to do.

It looks like you did a great job overall following one sentence per line, but there was just one line missing a couple line breaks in item 6. Also, I noticed that (probably due to Google Docs) the standard programming quotes were replaced with typographical "smart" quotes; we normally have a check for that but I had to disable it for a reason I can't recall. It would make the review comments a lot easier to follow if those little issues were cleaned up first; I can take care of it for you in a commit, or you can do it if you prefer—just let me know.

Thanks!

**Comment by hyounes4560** [comment]
@CAM-Gerlach thanks for your prompt response and helpful info. Yes, this PR is still a Draft as I am planning to revisit and expand other sections of the Editor's docs. It would make more sense if you guys did the review incrementally as I commit. However, it might be wiser to finish everything and mark the PR as ready. For now, you don't need to do anything.
Just for your info, I had a hard time pushing my commits since Thursday due to Pre-commits and had to bypass that using --no-verify. 
Thanks again.

**Comment by CAM-Gerlach** [comment]
> Yes, this PR is still a Draft as I am planning to revisit and expand other sections of the Editor's docs.

You could do the whole Editor docs in one PR, but especially with the size of the Editor pane, and the fact that its usually considered a good idea to keep PRs relatively focused on one specific thing, it might be a better idea to do it incrementally section by section, which each section corresponding to one PR (particularly given you've already aptly titled this one "Update the Editor introduction section", which I presumed was going to be the scope of the PR). This keeps the PRs more manageable, easier for us to review and easier for you to read the feedback, your changes get merged and go live quicker, and you can apply the feedback from one more easily to the next.

When you get a chance, could you let me know how you would like to approach

> When it comes to things like reST syntax, Sphinx roles, source formatting, and other technical and style guide details, there's a few ways we can handle that, so let me know which you prefer (

and

> It would make the review comments a lot easier to follow if those little issues were cleaned up first; I can take care of it for you in a commit, or you can do it if you prefer—just let me know.

Thanks!

> Just for your info, I had a hard time pushing my commits since Thursday due to Pre-commits and had to bypass that using --no-verify.

Thanks for letting me know. Since pre-commit is passing just fine on this PR as well as on my machine, it seems there is either an issue with your local setup, or a platform, etc. compatibility issue with one of the hooks. Either way it is something we need to fix, so could you open a new issue with some basic information about the problem so I can determine what is going on and either advise you how to fix it, or disable/revert the offending hook(s)? Specifically, could you please provide the following info:

* A brief description of the specific problem
* Full and exact invocation and output from a `pre-commit` run where you are seeing the problem (paste into a code block)
* Output of ``conda list`` (or `pip list`, if a virtualenv) of the environment you're running pre-commit in (paste into a code block)
* ``pre-commit --version``
* Your operating system name and version
* ``python -c "import sys; print(sys.version)"`` and what distribution (Python.org, Anaconda, Homebrew, distro, etc) you're using

Thanks!

**Comment by steff456** [comment]
> You could do the whole Editor docs in one PR, but especially with the size of the Editor pane, and the fact that its usually considered a good idea to keep PRs relatively focused on one specific thing, it might be a better idea to do it incrementally section by section, which each section corresponding to one PR (particularly given you've already aptly titled this one "Update the Editor introduction section", which I presumed was going to be the scope of the PR). This keeps the PRs more manageable, easier for us to review and easier for you to read the feedback, your changes get merged and go live quicker, and you can apply the feedback from one more easily to the next.

I think I also agree with this, so maybe what we can do is just open a PR for each section of the Editor docs. You will need to create a new branch from this PR to continue working on the following sections and then rebase when this one is merged. Please let us know if you need any help with this.

**Comment by hyounes4560** [comment]
> I see I was tagged for review, but this PR is also still marked as a draft. Just to confirm, would you like me to review it now? If so, is there anything particular you'd like to be in or out of scope for the first round of review?
> 
**It will mark as ready for review; so feel free to leave your feedback when applicable. Thanks**

> When it comes to things like reST syntax, Sphinx roles, source formatting, and other technical and style guide details, there's a few ways we can handle that, so let me know which you prefer (and we can always change it later on future PRs depending on what works best for you):
> 
**Please take a look at my next comment.** 

> * Review suggestions on the PR, and you apply them
> * Push a commit to the PR taking care of technical nits (all substantive content feedback will still be made as suggestions)
> * Ignore these minor defects on this PR completely and merge with them present, and make an immediate followup PR to address them
> 
> In the past, working with our previous writer, I initially did review suggestions, but we subsequently found directly pushing a commit taking care of the small nitpicky stuff to be a lot more efficient for both of us and did that from then on, so that might be the best option, but its ultimately up to you what you want me to do. Please let me know!
> 
> It looks like you did a great job overall following one sentence per line, but there was just one line missing a couple line breaks in item 6. Also, I noticed that (probably due to Google Docs) the standard programming quotes were replaced with typographical "smart" quotes; we normally have a pre-commit hook that autofixes them but I had to disable it temporarily due to an issue. It would make the review comments a lot easier to follow if those little issues were cleaned up first; I can take care of it for you in a commit, or you can do it if you prefer—just let me know.
> 
**Fixed both issues and the Pre-commit hook works fine now on my machine. Thanks** 
> Thanks!

**Comment by hyounes4560** [comment]
> it might be a better idea to do it incrementally section by section, which each section corresponding to one PR (particularly given you've already aptly titled this one "Update the Editor introduction section", which I presumed was going to be the scope of the PR). This keeps the PRs more manageable, easier for us to review and easier for you to read the feedback, your changes get merged and go live quicker, and you can apply the feedback from one more easily to the next.
> 
**This works perfectly for me and that's the plan moving forward. Thanks**

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:25]: I like the idea here and I think this is good overall.

It did take me some time to parse the error message and the code and connect the two together here, even as an experienced Python and Spyder user, and I suspect the reason for the error message would not be obvious to many less experienced readers. I suggest picking a clearer error message and with simpler code to make it easier for newer users to see how the error message can tell them what the problem is.

Also, its a little hard to read due to the error message overlapping the error line, and most of the right hand side of the image could be cropped off to focus specifically on the important content.

Putting it all together, here's one example of what you could do:

image

In case you actually want to use it, here's the file (which was around 10x smaller than the original after optimization): [editor-code-error.zip](https://github.com/spyder-ide/spyder-docs/files/9256819/editor-code-error.zip)

Also, renaming as above to elide "pane":

```suggestion
.. image:: /images/editor/editor-code-error.png
```

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:26]:
```suggestion
   :alt: Spyder Editor, showing an example of a code error popup message
```

This is still the Editor; the Outline is a separate pane. Also, be a little more descriptive and specific about what's being shown

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:34]:
```suggestion
```

At least per the current convention used on all the other panes that have substantial Options menus, we have a separate section specifically describing the functionality, where this screenshot would likely be a better fit, since it will be needed there anyway. Therefore, I would suggest saving it for that section and linking it here once you add it, which would also avoid it breaking up the flow of the numbered list of components and would make them shorter, where it is easier for readers to see both the image and the annotations and thus easier to understand and follow.

Also, for when this is used, it should be cropped to focus just on the outline menu, like all the similar screenshots in the other panes' docs are. This takes up much less space on screen and in the file, helps the reader focus on what's important and makes it much easier to read; as it stands now, I find it very hard to read the text.

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Thanks so much for your hard work on this, @hyounes4560 ! Here's a first pass of the review, focusing on the images and their alt text. I'll have a followup review focusing on the textual content to keep things organized, once you let me know how you want me to do that. Thanks!

For the images, to both load faster and more efficiently for readers and reduce the space they take up on this repo (forever), either you or I (let me know which you prefer) can reduce them to 256 colors and run [OptiPNG](http://optipng.sourceforge.net/) on them before merging. Together with the cropping I've suggested, this reduces their size by around 4-5x, without losing any perceptible quality.

To avoid multiple copies of the image files getting committed (forever) to this repo, I can just do a squash merge when merging (which doesn't touch your branch, you don't have to worry as much about your commit history and is a lot less work than doing a big rebase at the end).

File: doc/panes/editor.rst
Comment (hyounes4560) [reply doc/panes/editor.rst:18]: Will use it; thanks

File: doc/panes/editor.rst
Comment (hyounes4560) [reply doc/panes/editor.rst:18]: Will leave for now as I will be writing about code analysis and may need this image. tnx

File: doc/panes/editor.rst
Comment (hyounes4560) [reply doc/panes/editor.rst:25]: The purpose here is to show users any error/warning in their code; it's not about the clarity of the message nor the complexity of the code. This was just an example error message and in their own context, error/warning messages would make more sense to them. anyway; I will use your image. Thanks

Comment (hyounes4560) [reply doc/panes/editor.rst:25]: I have a question - why are we keeping this image? As per your last review comment you asked to remove the option menu's image **"which would also avoid it breaking up the flow of the numbered list of components and would make them shorter, where it is easier for readers to see both the image and the annotations and thus easier to understand and follow."**

**Comment by hyounes4560** [comment]
> I'll have a followup review focusing on the textual content to keep things organized, once you let me know how you want me to do that. Thanks!
> 
**I am not sure what you mean here! do what exactly?**

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [reply doc/panes/editor.rst:18]: Okay, sounds good.

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [reply doc/panes/editor.rst]: > The purpose here is to show users any error/warning in their code; it's not about the clarity of the message nor the complexity of the code.

Yeah, that's true, the code and the specific message isn't so important; we just don't want them to distract or confuse users or detract from the focus on what the screenshot *is* trying to communicate.

> I have a question - why are we keeping this image?

That is a good question, and one I'd also had myself but wasn't totally sure about the answer. The nominal rationale for keeping it unlike the other is that it is a much smaller image, wouldn't be duplicated somewhere else and doesn't require additional work. But if you're already going to have a section discussing the code analysis features, you could just use this image (or others of your choice) over there, which might be better. I defer to your judgement here.

If you do, make sure to remove the added image from this PR; it'll still be available on older commits in your branch and you can also keep a copy somewhere else convenient, of course.

**Comment by CAM-Gerlach** [comment]
Hey, apologies for the delay and I appreciate your patience; I somehow missed in my notifications that this PR had been updated, and only noticed when I went to check on it manually. I replied to your questions above and the text pass is in work right now.

> I am not sure what you mean here! do what exactly?

Sorry for the confusion; that was just the question about whether you preferred those smaller technical items as suggestions, a commit or a separate PR, to which you already let me know over Slack that you'd prefer suggestions for now, which I'll now go ahead with—I didn't want to have too many at once with the image suggestions also here.

Also, one other option for the future I didn't think of—I can also make a PR to your branch which you can then review and potentially incorporate as desired. That's just one other option to have of course; I'll do the suggestions here and going forward unless and until you let me know otherwise. Thanks!

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [reply doc/panes/editor.rst]: By the way, for this it would have been really helpful to have a high-level outline of the sections you plan to write and the overall content of each. However, I can't seem to find these documents in the Documentation folder of our Spyder Google Drive, and I don't believe I've received a link to them. If you have something like that, do you think either you or @steff456 could check on that? Thanks!

**Review by steff456 — APPROVED** [review]
Thanks for working on this @hyounes4560, I think this is ready!

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:22]:
```suggestion
   Clicking a line number selects the code on that line.
```

I actually never knew this, I always just triple-clicked the line—learned a new thing!

* Use clearer and more accurate terminology ("highlight" refers to merely visual _display_, e.g. that the line is shown in a brighter color when clicked, which is what I thought this meant at first, whereas "select" refers to the actual UI _functionality_ of text selection, as actually meant here)
* Use a slightly more idiomatic conjunction

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:27]:
```suggestion
2. The scrollbars allow vertical and horizontal navigation in a file.
```

* Use clearer, easier to read/understand and more comfortable structure
* Be more concise, particularly since a scrollbar is common to almost all graphical applications, and this section is a fairly high-level summary

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:28]:
```suggestion
3. The context (right-click) menu displays actions relevant to whatever was clicked.
```

* The context menu changes depending on where in the Editor (or indeed, anywhere in the UI) you click, so clarify that here
* Move "right-click" from previous sentence
* Fix "Context" capitalization

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:29]:
```suggestion
```

Since a context menu is a common feature of most Spyder panes, and indeed many/most graphical applications, I don't think we need to go into this level of detail here, especially since this is supposed to be a fairly high-level summary, and it repeats a lot of the previous sentence. Instead, we can just move the key bit, "right-click", to the previous sentence, and elide the rest.

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:30]:
```suggestion
4. The options menu ("Hamburger" icon at top right) includes useful settings and actions relevant to the Editor.
```

* Revise phrasing to be simpler, more concise and easier to read
* The others all lead with "The [thing name] [what it does]", so frame this the same way
* The "options menu" is lowercase, by convention (we don't use Proper Case unless its actually named such in the GUI), while Editor is Proper Case

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Text looks great overall; thanks @hyounes4560 ! I just had some mostly minor suggestions you can apply.

Once that's done, it should be all good to merge, aside from making sure to delete from this branch the options menu screenshot file added here which is no longer used in this PR, as well as the error message screenshot if you decide to defer that to a later PR too, per [our discussion](https://github.com/spyder-ide/spyder-docs/pull/322#discussion_r937350224)). Thanks!

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:7]:
```suggestion
In addition, it integrates a number of powerful tools for an easy to use, efficient editing experience.
```

Teeny tiny nit, but It looks like a stray space was retained when this change was manually applied—as a tip, if you use GitHub's "Apply suggestions" feature as I've mentioned, you can be sure that any mistakes are 100% on me :joy: (plus, its a lot faster and easier both for reviewers and yourself :)

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:22]:
```suggestion
   :alt: Spyder's Editor pane, showing its different areas (described below)

1. The left sidebar shows line numbers and displays any code analysis errors/warnings that exist in the current file.
```

It looks like relative to my suggestions, an extra line break got added here...

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:31]:
```suggestion
   It also has a :guilabel:`Browse tabs` button (at left) to show every open tab and switch between them—which comes in handy if many are open.



===================
```

And removed here

**Review by CAM-Gerlach — APPROVED** [review]
Just a couple trivial spacing issues that can be applied as suggestions (`Files` -> `Add to batch` -> `Commit`) and this LGTM and looks ready to be merged! Thanks, @hyounes4560 !

**Review by CAM-Gerlach — APPROVED** [review]
This looks good to me @hyounes4560 ! Thanks again for your efforts. I will ask @ccordoba12 if he wants to give this a quick look; otherwise we should be good to merge and see your work go live for everyone!

**Comment by hyounes4560** [comment]
FYI @ccordoba12 Thanks

**Comment by CAM-Gerlach** [comment]
I brought this up at the meeting today, and @ccordoba12 said he will review it today or tomorrow, FYI

File: doc/panes/editor.rst
Comment (ccordoba12) [inline doc/panes/editor.rst:22]: Should we mention here that users can set breakpoints by doing a right-click to the right of line numbers?

**Review by ccordoba12 — COMMENTED** [review]
Hey @hyounes4560, great work here! I really like the screenshot with numbers for different areas in the editor. It's a really neat way of introducing the most important functionalities available in that pane.

I only left a minor comment for you and the other reviewers. The rest looks good to me.

File: doc/panes/editor.rst
Comment (CAM-Gerlach) [inline doc/panes/editor.rst:22]: What about this?

```suggestion
1. The left sidebar shows line numbers and displays any code analysis warnings that exist in the current file.
   Left-clicking a line number selects the text on that line, and right-clicking to the right of it sets a :ref:`breakpoint <debugging-breakpoints>`.
```

You'll also need to add an appropriately named reference target in `panes/debugging.rst` like so (FYI, I really should go in and add ref targets to all headings, so we don't need to do this in the future):

```rst
.. image:: /images/debugging/debugging-multiline.gif
   :alt: Debugger showing multiline debugger



.. _debugging-breakpoints:

===========
Breakpoints
===========
```

I also simplified it a touch elsewhere to compensate.

Comment (jitseniesen) [reply doc/panes/editor.rst:22]: Isn't it **left**-clicking to the right of the line number to set a breakpoint?

Comment (ccordoba12) [reply doc/panes/editor.rst:22]: You're totally right @jitseniesen! Thanks for catching that.

Comment (ccordoba12) [reply doc/panes/editor.rst:22]: My suggestion

```suggestion
1. The left sidebar shows line numbers and displays any code analysis errors/warnings that exist in the current file.
   Left-clicking a line number selects the text on that line, and doing it to the right of it sets a :ref:`breakpoint <debugging-breakpoints>`.
```

My suggestion to address @jitseniesen's comment.

Comment (CAM-Gerlach) [reply doc/panes/editor.rst:22]: Thanks, I should have noticed that too, sorry. I've edited the original, and repeat it here for clarity:

```suggestion
1. The left sidebar shows line numbers and displays any code analysis warnings that exist in the current file.
   Clicking a line number selects the text on that line, and clicking to the right of it sets a :ref:`breakpoint <debugging-breakpoints>`.
```

Comment (CAM-Gerlach) [reply doc/panes/editor.rst:22]: (See discussion below)

Comment (ccordoba12) [reply doc/panes/editor.rst:22]: Ok, I agree with this suggestion.

Comment (CAM-Gerlach) [reply doc/panes/editor.rst:22]: Just so it doesn't get lost, @hyounes4560 , in addition to the suggestion, you'll also need to add the indicated ref target label to `panes/debugging.rst` as [shown above](https://github.com/spyder-ide/spyder-docs/pull/322#discussion_r944763545). Thanks!

**Review by ccordoba12 — APPROVED** [review]
Thanks @hyounes4560!

**Review by jitseniesen — APPROVED** [review]
Looks good, well done!

**Comment by CAM-Gerlach** [comment]
Thanks to @hyounes4560 for the PR, @steff456 for helping guide her and @ccordoba12 and @jitseniesen for the reviews! This is a big improvement, and looking forward to the next one!

**Pull request merged**
